File: pdf2cho.js
/* PDF -> ChordPro (.cho) conversion, in the browser.
 *
 * A port of tools/pdf2cho.py: reconstruct a monospace layout from word x/y
 * boxes, then merge each chord line into the lyric line below it as inline
 * [chord] tokens.
 *
 * Text comes from the PDF's own text layer when it has one (exact and instant).
 * Image-only PDFs -- the Ultimate Guitar exports this was built for -- fall back
 * to OCR. pdf.js and Tesseract are fetched from a CDN on first use only, so the
 * app itself stays dependency-free and costs nothing to load for everyone who
 * never drops a PDF.
 */
(function (root) {
  'use strict';

  const PDFJS = 'https://cdn.jsdelivr.net/npm/pdfjs-dist@4.6.82/build/pdf.min.mjs';
  const PDFJS_WORKER = 'https://cdn.jsdelivr.net/npm/pdfjs-dist@4.6.82/build/pdf.worker.min.mjs';
  const TESS = 'https://cdn.jsdelivr.net/npm/tesseract.js@5.1.1/dist/tesseract.min.js';
  const TESS_WORKER = 'https://cdn.jsdelivr.net/npm/tesseract.js@5.1.1/dist/worker.min.js';
  const TESS_CORE = 'https://cdn.jsdelivr.net/npm/tesseract.js-core@5.1.0';
  const TESS_LANG = 'https://cdn.jsdelivr.net/gh/naptha/tessdata@gh-pages/4.0.0_fast';

  const OCR_DPI = 200;          // what the Python pipeline renders at
  const TEXT_CHARS_PER_PAGE = 40; // below this, treat the PDF as image-only

  // ---- chord grammar (mirrors pdf2cho.py) ----------------------------------

  const QUAL =
    '(?:maj7|maj9|maj|min7|min|sus2|sus4|sus|add9|add11|add2|add|aug|' +
    'dim7|dim|m7b5|m7|m9|m6|m11|m|°|Δ|6|7|9|11|13|2|4|5|#11|#9|b9|#5|b5)';
  // Sticky, so exec() at lastIndex anchors there -- Python's _ONE.match(tok, i).
  const ONE = new RegExp('[A-Ga-g][#b]?' + QUAL + '*(?:/[A-Ga-g][#b]?)?', 'y');

  // An uppercase chord blob = an instrumental row whose spaces OCR dropped:
  // "GBC", "EAEA", "DGDAG", "D5G5".
  const INSTR_BLOB = /^[A-G][A-G0-9#b/]*$/;
  const SECTION_RE = /^\[([A-Za-z][\w /'-]{0,24})\]$/;
  const HEADER_RE = /https?:\/\/|chordify|^\s*(capo|tuning|difficulty|tabbed by|artist|song|key)\b/i;

  function ocrFix(tok) {
    // Narrow, safe OCR corrections. Both require an uppercase root, so common
    // lowercase words ('an', 'as', 'son') are untouched.
    tok = tok.replace(/(?<=[A-G])S\b/g, '5');        // power chord: DS -> D5
    tok = tok.replace(/(?<=[A-G])n([79]?)$/, 'm$1'); // minor: Dn -> Dm
    return tok;
  }

  function normOne(c) {
    c = c[0].toUpperCase() + c.slice(1);
    return c.replace(/\/([a-g])/, (_, g) => '/' + g.toUpperCase());
  }

  // If the whole token is a run of concatenated chords, return them as a list
  // (splitting OCR-merged blobs like 'EmD' -> ['Em','D']); else null.
  function parseChords(tok, allowBare) {
    tok = ocrFix(String(tok).replace(/\|/g, ''));
    if (!tok) return null;
    const out = [];
    let i = 0;
    while (i < tok.length) {
      ONE.lastIndex = i;
      const m = ONE.exec(tok);
      if (!m || !m[0]) return null;
      out.push(normOne(m[0]));
      i = ONE.lastIndex;
    }
    const dd = []; // collapse a chord doubled within one blob (OCR echo: 'Cc')
    for (const c of out) if (!dd.length || dd[dd.length - 1] !== c) dd.push(c);
    // 3+ bare roots is ambiguous: a lyric word ("bad") or a chord row ("FCGF").
    // Case decides -- words are lowercase -- but only trust that in chord context.
    if (dd.length >= 3 && dd.every((c) => /^[A-G]$/.test(c))) {
      if (!(allowBare && !/[a-z]/.test(tok))) return null;
    }
    return dd.length ? dd : null;
  }

  function isChordLine(text) {
    const toks = text.replace(/\|/g, ' ').split(/\s+/).filter(Boolean);
    if (!toks.length) return false;
    const good = toks.filter((t) => parseChords(t)).length;
    return good >= 1 && good / toks.length >= 0.6;
  }

  const collapse = (s) => s.replace(/ {2,}/g, ' ').trim();

  // Insert [chord] into the lyric at each chord's column, then tidy spacing.
  // An OCR-merged blob expands to several chords sharing one column, since their
  // true separate columns are unrecoverable.
  function merge(chordText, lyric) {
    const inserts = [];
    let seq = 0;
    const re = /\S+/g;
    let m;
    while ((m = re.exec(chordText.replace(/\|/g, ' ')))) {
      const chords = parseChords(m[0], true);
      if (!chords) continue;
      for (const ch of chords) inserts.push([m.index, seq++, '[' + ch + ']']);
    }
    inserts.sort((a, b) => b[0] - a[0] || b[1] - a[1]);
    let s = lyric;
    for (const ins of inserts) {
      const c = Math.min(ins[0], s.length);
      s = s.slice(0, c) + ins[2] + s.slice(c);
    }
    return collapse(s.replace(/(\]) +/g, '$1')); // keep [chord] glued to its word
  }

  function bracketChordLine(text) {
    const out = [];
    for (const t of text.replace(/\|/g, ' ').split(/\s+/).filter(Boolean)) {
      const chords = parseChords(t, true);
      if (chords) out.push(...chords.map((c) => '[' + c + ']'));
      else out.push(t);
    }
    return out.join(' ');
  }

  // ---- layout reconstruction ------------------------------------------------

  function median(xs) {
    if (!xs.length) return 0;
    const a = [...xs].sort((p, q) => p - q);
    const m = a.length >> 1;
    return a.length % 2 ? a[m] : (a[m - 1] + a[m]) / 2;
  }

  // Group words into visual lines by vertical overlap; sort each by x.
  function groupLines(words) {
    const ws = [...words].sort((a, b) => a.top - b.top || a.left - b.left);
    const h = ws.length ? median(ws.map((w) => w.height)) : 12;
    const lines = [];
    let cur = [], curTop = null;
    for (const w of ws) {
      if (curTop === null || Math.abs(w.top - curTop) <= h * 0.6) {
        cur.push(w);
        curTop = curTop === null ? w.top : (curTop + w.top) / 2;
      } else {
        lines.push(cur.sort((a, b) => a.left - b.left));
        cur = [w];
        curTop = w.top;
      }
    }
    if (cur.length) lines.push(cur.sort((a, b) => a.left - b.left));
    return lines;
  }

  // Estimate monospace char width from word width / length.
  function charPx(words) {
    const r = words.filter((w) => w.text.length >= 3).map((w) => w.width / w.text.length);
    return r.length ? median(r) : 10;
  }

  function layoutLine(line, cpx, x0) {
    let s = '';
    for (const w of line) {
      let col = Math.max(0, Math.round((w.left - x0) / cpx));
      if (col < s.length) col = s.length + 1;
      s += ' '.repeat(col - s.length) + w.text;
    }
    return s;
  }

  function linesFromWords(words) {
    if (!words.length) return [];
    const cpx = charPx(words);
    let x0 = Infinity;
    for (const w of words) if (w.left < x0) x0 = w.left; // reduce, not Math.min(...)
    return groupLines(words).map((ln) => layoutLine(ln, cpx, x0));
  }

  // ---- .cho assembly --------------------------------------------------------

  function parseMeta(lines) {
    const meta = {};
    for (const ln of lines.slice(0, 20)) {
      const t = ln.trim();
      const m = t.match(/^(.*?)\s+Chords\s+by\s+(.+?)(?:\s+[a-z]{1,3})?$/i);
      if (m && !meta.title) {
        meta.title = collapse(m[1]);
        meta.artist = collapse(m[2].replace(/\s+[a-z]{1,3}$/, ''));
      }
      const mc = t.match(/^Capo:?\s*(\d+)/i);
      if (mc) meta.capo = mc[1];
      const mk = t.match(/^Key:?\s*([A-G][#b]?m?)\b/i);
      if (mk) meta.key = mk[1];
    }
    return meta;
  }

  // Where the song body begins. Normally the first {Section}; charts without
  // section labels (e.g. chordify exports) need the header skipped and the body
  // started at the first chord line.
  function findBodyStart(lines) {
    const sec = lines.findIndex((ln) => SECTION_RE.test(ln.trim()));
    if (sec !== -1) return sec;
    for (let i = 0; i < lines.length; i++) {
      const s = lines[i].trim();
      if (i === 0 || !s || HEADER_RE.test(s)) continue; // skip title, blanks, notes
      if (isChordLine(s)) return i;
    }
    for (let i = 0; i < lines.length; i++) { // fallback: first real content line
      const s = lines[i].trim();
      if (i === 0 || !s || HEADER_RE.test(s)) continue;
      return i;
    }
    return lines.length;
  }

  function toCho(lines, title) {
    const meta = parseMeta(lines);
    // The filename is the clean song title; OCR of the header is noisy
    // ("I Won't Back Down" -> "| Wont Back Down"), so always prefer it.
    meta.title = title;
    const bodySrc = lines
      .slice(findBodyStart(lines))
      .map((l) => l.replace(/\s+$/, ''))
      .filter((l) => !/^\s*Page\s*\d/.test(l));

    const out = [];
    let i = 0;
    while (i < bodySrc.length) {
      const ln = bodySrc[i];
      const s = ln.trim();
      const sec = s.match(SECTION_RE);
      if (sec) { out.push('{' + collapse(sec[1]) + '}'); i++; continue; }
      if (!s) { out.push(''); i++; continue; }
      if (isChordLine(s)) {
        const nxt = i + 1 < bodySrc.length ? bodySrc[i + 1] : '';
        const ns = nxt.trim();
        // Merge into the next line only if it looks like real lyrics (several
        // words) -- avoids corrupting a chord row whose OCR-mangled neighbour is
        // a single gibberish token.
        if (ns && ns.includes(' ') && !isChordLine(ns) && !SECTION_RE.test(ns)) {
          out.push(merge(ln, nxt)); i += 2; continue;
        }
        out.push(bracketChordLine(collapse(s))); i++; continue;
      }
      // A lone uppercase chord blob (spaces dropped by OCR) = instrumental row.
      if (!s.includes(' ') && INSTR_BLOB.test(s)) {
        const chords = parseChords(s, true);
        if (chords && chords.length >= 2) {
          out.push(chords.map((c) => '[' + c + ']').join(' ')); i++; continue;
        }
      }
      out.push(collapse(s)); i++;
    }

    let hdr = '';
    if (meta.title) hdr += '{title: ' + meta.title + '}\n';
    if (meta.artist) hdr += '{artist: ' + meta.artist + '}\n';
    if (meta.key) hdr += '{key: ' + meta.key + '}\n';
    if (meta.capo) hdr += '{capo: ' + meta.capo + '}\n';
    let body = out.join('\n').replace(/^\n+|\n+$/g, '');
    body = body.replace(/\n{3,}/g, '\n\n');
    return (hdr ? hdr + '\n' + body : body) + '\n';
  }

  // ---- cleanup (mirrors tools/cleanup.py) -----------------------------------

  const CHORD_ONLY = /^(?:\[[^\]]+\]\s*)+$/;
  const INTERJECTIONS = new Set(['oh', 'ah', 'ooh', 'oooh', 'whoa', 'woah', 'hey',
    'la', 'na', 'yeah', 'mmm', 'hmm', 'ooo', 'ohh', 'ahh', 'yo', 'uh']);

  // A smashed OCR echo of the chord row above ("GopeC", "CG", "Goce"). Guards
  // hard against real content: section labels, punctuated interjections and
  // common backing-vocal words must survive.
  function isGarbageEcho(tok, prevChordLine) {
    if (tok[0] === '{' || tok[0] === '[') return false;   // section label / chord
    if (/[!?.,;:'"()\-]/.test(tok)) return false;         // punctuation => lyric
    if (!/^[A-Za-z]+$/.test(tok)) return false;           // digits/symbols
    if (INTERJECTIONS.has(tok.toLowerCase())) return false;
    if (/[A-Z]/.test(tok.slice(1))) return true;          // interior caps (GopeC)
    if (tok === tok.toUpperCase() && tok.length <= 4) return true; // short all-caps
    if (!/[aeiou]/i.test(tok)) return true;               // no vowels (Gbpc)
    // Short title-case blob (Goce) -- only if it mirrors a chord root above, so
    // real one-word lyrics ("Come", "Down") aren't nuked unless the row echoes.
    const roots = new Set([...prevChordLine.matchAll(/\[([A-G])/g)].map((m) => m[1]));
    return tok.length <= 5 && roots.has(tok[0]);
  }

  function cleanCho(text) {
    const out = [];
    let prev = '';
    for (const line of text.split('\n')) {
      const s = line.trim();
      // Only drop a garbled echo that DUPLICATES a clean chord row right above
      // it (safe: the real chords are already captured). Never drop a lone blob
      // under a section label -- that's an instrumental row's only chord info.
      if (s && !s.includes(' ') && prev && CHORD_ONLY.test(prev) && isGarbageEcho(s, prev)) continue;
      out.push(line);
      if (s) prev = s;
    }
    return out.join('\n');
  }

  // ---- lazy CDN loaders -----------------------------------------------------

  let pdfjsP = null;
  function loadPdfjs() {
    if (!pdfjsP) {
      pdfjsP = import(/* webpackIgnore: true */ PDFJS).then((m) => {
        m.GlobalWorkerOptions.workerSrc = PDFJS_WORKER;
        return m;
      }).catch((e) => { pdfjsP = null; throw new Error('Could not load the PDF reader. ' + (navigator.onLine ? '' : 'You appear to be offline.')); });
    }
    return pdfjsP;
  }

  let tessP = null;
  function loadTesseract() {
    if (!tessP) {
      tessP = new Promise((res, rej) => {
        const sc = document.createElement('script');
        sc.src = TESS;
        sc.onload = () => res(root.Tesseract);
        sc.onerror = () => { tessP = null; rej(new Error('Could not load the OCR engine. ' + (navigator.onLine ? '' : 'You appear to be offline.'))); };
        document.head.appendChild(sc);
      });
    }
    return tessP;
  }

  // ---- page -> words --------------------------------------------------------

  async function wordsFromTextLayer(page, pdfjs) {
    const vp = page.getViewport({ scale: 1 });
    const tc = await page.getTextContent();
    const words = [];
    for (const it of tc.items) {
      const text = (it.str || '').trim();
      if (!text) continue;
      const t = pdfjs.Util.transform(vp.transform, it.transform);
      const h = Math.hypot(t[2], t[3]) || 10;
      words.push({
        text,
        left: t[4],
        top: t[5] - h, // transform lands on the baseline; lift to the box top
        width: it.width || text.length * h * 0.5,
        height: h,
      });
    }
    return words;
  }

  async function wordsFromOcr(page, worker) {
    const vp = page.getViewport({ scale: OCR_DPI / 72 });
    const canvas = document.createElement('canvas');
    canvas.width = Math.ceil(vp.width);
    canvas.height = Math.ceil(vp.height);
    await page.render({ canvasContext: canvas.getContext('2d'), viewport: vp }).promise;
    const { data } = await worker.recognize(canvas, {}, { blocks: true, text: false });
    const words = [];
    // Walk blocks rather than data.words: the flattened getter is version-shaky.
    for (const b of data.blocks || [])
      for (const p of b.paragraphs || [])
        for (const l of p.lines || [])
          for (const w of l.words || []) {
            const text = (w.text || '').trim();
            if (!text || !w.bbox) continue;
            words.push({
              text,
              left: w.bbox.x0,
              top: w.bbox.y0,
              width: w.bbox.x1 - w.bbox.x0,
              height: w.bbox.y1 - w.bbox.y0,
            });
          }
    return words;
  }

  // ---- entry point ----------------------------------------------------------

  /* Convert a PDF File/Blob to .cho text.
   * onProgress(message) is called with human-readable status along the way.
   * Resolves { cho, usedOcr, pages }. */
  async function pdfToCho(file, onProgress) {
    const say = onProgress || (() => {});
    say('Reading PDF…');
    const pdfjs = await loadPdfjs();
    const doc = await pdfjs.getDocument({ data: await file.arrayBuffer() }).promise;

    // Try the PDF's own text layer first -- exact, instant, and it skips the
    // multi-MB OCR download entirely for charts that carry real text.
    const pages = [];
    let chars = 0;
    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i);
      const words = await wordsFromTextLayer(page, pdfjs);
      chars += words.reduce((n, w) => n + w.text.length, 0);
      pages.push({ page, words });
    }

    const usedOcr = chars < TEXT_CHARS_PER_PAGE * doc.numPages;
    if (usedOcr) {
      say('Loading OCR engine… (one-time download)');
      const Tesseract = await loadTesseract();
      const worker = await Tesseract.createWorker('eng', 1, {
        workerPath: TESS_WORKER,
        corePath: TESS_CORE,
        langPath: TESS_LANG,
        logger: (m) => {
          if (m.status && m.status !== 'recognizing text') say(m.status.replace(/^\w/, (c) => c.toUpperCase()) + '…');
        },
      });
      // --psm 4: a single column of text of variable sizes, as the CLI used.
      await worker.setParameters({ tessedit_pageseg_mode: '4' });
      try {
        for (let i = 0; i < pages.length; i++) {
          say('Reading page ' + (i + 1) + ' of ' + pages.length + '…');
          pages[i].words = await wordsFromOcr(pages[i].page, worker);
        }
      } finally {
        await worker.terminate();
      }
    }

    say('Building chart…');
    const lines = [];
    for (const p of pages) lines.push(...linesFromWords(p.words));
    const title = file.name.replace(/\.[^.]+$/, '');
    return { cho: cleanCho(toCho(lines, title)), usedOcr, pages: doc.numPages };
  }

  root.PdfToCho = {
    pdfToCho,
    // exposed for the port-fidelity test against the Python pipeline
    _internals: { toCho, cleanCho, parseChords, isChordLine, merge, linesFromWords },
  };
})(typeof window !== 'undefined' ? window : globalThis);

File: tools/pdf2cho.py
#!/usr/bin/env python3
"""OCR an image-based Ultimate Guitar PDF into a .cho chords-over-lyrics file.

Pipeline: pdftoppm -> tesseract TSV (word boxes) -> reconstruct monospace layout
from x-positions -> merge chord lines into lyric lines as inline [chord].
"""
import subprocess, sys, re, os, tempfile, statistics

CHORD_RE = re.compile(r'^[A-G][#b]?(?:maj|min|sus|add|aug|dim|m|M|Δ|°|\+|-|[0-9]|#|b|\(|\))*(?:/[A-G][#b]?)?$')

# One chord anywhere in a string (not anchored) — root + optional accidental +
# optional quality/extensions + optional slash bass. Lowercase roots allowed
# because OCR frequently lowercases them.
_QUAL = (r'(?:maj7|maj9|maj|min7|min|sus2|sus4|sus|add9|add11|add2|add|aug|'
         r'dim7|dim|m7b5|m7|m9|m6|m11|m|°|Δ|6|7|9|11|13|2|4|5|#11|#9|b9|#5|b5)')
_ONE = re.compile(r'[A-Ga-g][#b]?' + _QUAL + r'*(?:/[A-Ga-g][#b]?)?')

def clean_chord_tok(t):
    # OCR often lowercases the root (C->c) or doubles it (C->Cc); fix both.
    if len(t) >= 2 and t[0].lower() == t[1].lower() and t[0].lower() in 'abcdefg':
        t = t[0] + t[2:]
    return t[0].upper() + t[1:] if t else t

def _norm_one(c):
    c = c[0].upper() + c[1:]
    return re.sub(r'/([a-g])', lambda m: '/' + m.group(1).upper(), c)

def _ocr_fix(tok):
    """Narrow, safe OCR corrections inside chord tokens. Both target an uppercase
    root so common lowercase words ('an', 'as', 'son') are untouched."""
    tok = re.sub(r'(?<=[A-G])S\b', '5', tok)       # power chord: DS -> D5
    tok = re.sub(r'(?<=[A-G])n([79]?)$', r'm\1', tok)  # minor: Dn -> Dm
    return tok

def parse_chords(tok, allow_bare=False):
    """If the whole token is a run of concatenated chords, return them as a list
    (splitting OCR-merged blobs like 'EmD' -> ['Em','D']); else None.
    allow_bare keeps runs of 3+ bare single-letter roots (real for instrumental
    rows like 'GBC', but a lyric word like 'bad' when reading lyric context)."""
    tok = _ocr_fix(tok.replace('|', ''))
    if not tok:
        return None
    out, i, n = [], 0, len(tok)
    while i < n:
        m = _ONE.match(tok, i)
        if not m or m.end() == i:
            return None
        out.append(_norm_one(m.group()))
        i = m.end()
    # collapse a chord doubled within one blob (OCR echo: 'Cc' -> C)
    dd = []
    for c in out:
        if not dd or dd[-1] != c:
            dd.append(c)
    # 3+ bare single-letter roots is ambiguous: a lyric word ("bad"->B,A,D) or an
    # instrumental chord row ("FCGF"). Case decides — words are lowercase, chord
    # rows uppercase — but only trust that in chord context (allow_bare).
    if len(dd) >= 3 and all(re.fullmatch(r'[A-G]', c) for c in dd):
        if not (allow_bare and not any(ch.islower() for ch in tok)):
            return None
    return dd or None

# An uppercase chord blob = an instrumental row (intro/solo/outro) whose spaces
# OCR dropped: "GBC", "EAEA", "DGDAG", "D5G5". All-uppercase distinguishes it from
# a lyric word ("bad", "Dad"), so bare 3+ root runs are safe to bracket here.
INSTR_BLOB = re.compile(r'^[A-G][A-G0-9#b/]*$')

def is_chord(tok):
    return parse_chords(tok) is not None

def render_pages(pdf, dpi=200):
    d = tempfile.mkdtemp()
    subprocess.run(['pdftoppm', '-png', '-r', str(dpi), pdf, os.path.join(d, 'p')],
                   check=True, stderr=subprocess.DEVNULL)
    return sorted(os.path.join(d, f) for f in os.listdir(d) if f.endswith('.png'))

def ocr_words(png):
    """Return list of words: {text,left,top,width,height,conf,line_key}."""
    out = subprocess.run(['tesseract', png, '-', '--psm', '4', 'tsv'],
                         capture_output=True, text=True).stdout
    words = []
    for row in out.splitlines()[1:]:
        c = row.split('\t')
        if len(c) < 12 or c[0] != '5':
            continue
        text = c[11].strip()
        if not text:
            continue
        words.append(dict(text=text, left=int(c[6]), top=int(c[7]), width=int(c[8]),
                          height=int(c[9]), conf=float(c[10]),
                          line_key=(int(c[2]), int(c[3]), int(c[4]))))
    return words

def group_lines(words):
    """Group words into visual lines by vertical overlap; sort each by x."""
    words = sorted(words, key=lambda w: (w['top'], w['left']))
    lines, cur, cur_top, h = [], [], None, statistics.median([w['height'] for w in words]) if words else 12
    for w in words:
        if cur_top is None or abs(w['top'] - cur_top) <= h * 0.6:
            cur.append(w); cur_top = w['top'] if cur_top is None else (cur_top + w['top']) / 2
        else:
            lines.append(sorted(cur, key=lambda x: x['left'])); cur = [w]; cur_top = w['top']
    if cur:
        lines.append(sorted(cur, key=lambda x: x['left']))
    return lines

def char_px(words):
    """Estimate monospace char width from word width / length."""
    ratios = [w['width'] / len(w['text']) for w in words if len(w['text']) >= 3]
    return statistics.median(ratios) if ratios else 10.0

def layout_line(line, cpx, x0):
    """Render one visual line as monospace text using x-positions."""
    s = ''
    for w in line:
        col = max(0, round((w['left'] - x0) / cpx))
        if col < len(s):
            col = len(s) + 1
        s += ' ' * (col - len(s)) + w['text']
    return s

def is_chord_line(text):
    toks = text.replace('|', ' ').split()
    if not toks:
        return False
    good = sum(1 for t in toks if parse_chords(t))
    return good >= 1 and good / len(toks) >= 0.6

def collapse(s):
    return re.sub(r' {2,}', ' ', s).strip()

def merge(chord_text, lyric):
    """Insert [chord] into lyric at each chord's column.
    OCR-merged blobs (e.g. 'EmD') expand to several chords placed adjacently,
    since their true separate columns are unrecoverable."""
    inserts, seq = [], 0
    for m in re.finditer(r'\S+', chord_text.replace('|', ' ')):
        chords = parse_chords(m.group(), allow_bare=True)
        if not chords:
            continue
        # A chord whose column falls in the gap between two words belongs to the
        # next word -- slide past the spaces to its start. Inserting on the gap
        # instead would consume it and smash the words together ("one[G]word").
        col = m.start()
        while col < len(lyric) and lyric[col] == ' ':
            col += 1
        # Clamp overhanging chords to the lyric end NOW, not at apply time: the
        # string grows as brackets go in, so a late clamp lands each chord at a
        # different spot -- reversing a trailing turnaround and splitting
        # brackets. Clamped to one column, they share it and seq keeps order.
        col = min(col, len(lyric))
        for ch in chords:  # merged blob -> chords share the column, stay grouped
            inserts.append((col, seq, '[' + ch + ']')); seq += 1
    s = lyric
    for col, _, tag in sorted(inserts, key=lambda x: (x[0], x[1]), reverse=True):
        s = s[:min(col, len(s))] + tag + s[min(col, len(s)):]
    return collapse(s)  # chords now sit at word starts; no space-eating needed

SECTION_RE = re.compile(r'^\[([A-Za-z][\w /\'-]{0,24})\]$')

def normalize_chord(tok):
    return clean_chord_tok(tok)

def bracket_chord_line(text):
    out = []
    for t in text.replace('|', ' ').split():
        chords = parse_chords(t, allow_bare=True)
        if chords:
            out += ['[' + c + ']' for c in chords]
        else:
            out.append(t)
    return ' '.join(out)

def raw_lines(pdf):
    out = []
    for png in render_pages(pdf):
        words = ocr_words(png)
        if not words:
            continue
        cpx = char_px(words)
        x0 = min(w['left'] for w in words)
        for ln in group_lines(words):
            out.append(layout_line(ln, cpx, x0))
    return out

def parse_meta(lines):
    meta = {}
    for ln in lines[:20]:
        t = ln.strip()
        m = re.match(r'^(.*?)\s+Chords\s+by\s+(.+?)(?:\s+[a-z]{1,3})?$', t, re.I)
        if m and 'title' not in meta:
            meta['title'] = collapse(m.group(1))
            meta['artist'] = collapse(re.sub(r'\s+[a-z]{1,3}$', '', m.group(2)))
        mc = re.match(r'^Capo:?\s*(\d+)', t, re.I)
        if mc:
            meta['capo'] = int(mc.group(1))
        mk = re.match(r'^Key:?\s*([A-G][#b]?m?)\b', t, re.I)
        if mk:
            meta['key'] = mk.group(1)
    return meta

HEADER_RE = re.compile(r'https?://|chordify|^\s*(capo|tuning|difficulty|tabbed by|'
                       r'artist|song|key)\b', re.I)

def find_body_start(lines):
    """Where the song body begins. Normally the first {Section}; but charts
    without section labels (e.g. chordify exports) need the header skipped and
    the body started at the first chord line."""
    sec = next((i for i, ln in enumerate(lines) if SECTION_RE.match(ln.strip())), None)
    if sec is not None:
        return sec
    for i, ln in enumerate(lines):
        s = ln.strip()
        if i == 0 or not s or HEADER_RE.search(s):  # skip title, blanks, header notes
            continue
        if is_chord_line(s):
            return i
    for i, ln in enumerate(lines):                    # fallback: first real content line
        s = ln.strip()
        if i == 0 or not s or HEADER_RE.search(s):
            continue
        return i
    return len(lines)

def to_cho(pdf):
    lines = raw_lines(pdf)
    meta = parse_meta(lines)
    # The filename is the clean song title; OCR of the header is noisy
    # ("I Won't Back Down" -> "| Wont Back Down"), so always prefer the filename.
    meta['title'] = os.path.splitext(os.path.basename(pdf))[0]
    start = find_body_start(lines)
    body_src = [ln.rstrip() for ln in lines[start:] if not re.match(r'^\s*Page\s*\d', ln)]

    out = []
    i = 0
    while i < len(body_src):
        ln = body_src[i]
        s = ln.strip()
        sec = SECTION_RE.match(s)
        if sec:
            out.append('{' + collapse(sec.group(1)) + '}'); i += 1; continue
        if not s:
            out.append(''); i += 1; continue
        if is_chord_line(s):
            nxt = body_src[i + 1] if i + 1 < len(body_src) else ''
            ns = nxt.strip()
            # Merge into the next line only if it looks like real lyrics (has
            # several words) — avoids corrupting chord-only rows whose OCR-mangled
            # neighbour is a single gibberish token.
            if ns and ' ' in ns and not is_chord_line(ns) and not SECTION_RE.match(ns):
                out.append(merge(ln, nxt)); i += 2; continue
            out.append(bracket_chord_line(collapse(s))); i += 1; continue
        # A lone uppercase chord blob (spaces dropped by OCR) = instrumental row.
        if ' ' not in s and INSTR_BLOB.match(s):
            chords = parse_chords(s, allow_bare=True)
            if chords and len(chords) >= 2:
                out.append(' '.join('[' + c + ']' for c in chords)); i += 1; continue
        out.append(collapse(s)); i += 1

    hdr = ''
    if meta.get('title'): hdr += '{title: %s}\n' % meta['title']
    if meta.get('artist'): hdr += '{artist: %s}\n' % meta['artist']
    if meta.get('key'): hdr += '{key: %s}\n' % meta['key']
    if meta.get('capo'): hdr += '{capo: %s}\n' % meta['capo']
    body = '\n'.join(out).strip('\n')
    # collapse 3+ blank lines
    body = re.sub(r'\n{3,}', '\n\n', body)
    return (hdr + '\n' + body if hdr else body) + '\n'

if __name__ == '__main__':
    sys.stdout.write(to_cho(sys.argv[1]))
